### PR TITLE
fix(global virtual store): bridge the resolution paths a store slot lost

### LINKED ISSUE #10588: Global virtual store: hoisted-resolution bridge misses the workspace root's node_modules
## Summary

Under pnpm's global virtual store (`enableGlobalVirtualStore: true`), the hoisted-resolution bridge restores only half of the resolution surface a package had under the project-local virtual store. A package living in a store slot can still not reach the workspace's own `node_modules`, so any phantom require of a **direct** dependency of the workspace root fails.

## Background

`ensureHoistedDependencyResolution` ([hoisted-resolution-bridge.ts:205](https://github.com/teambit/bit/blob/master/scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts#L205)) puts one directory on `NODE_PATH` (plus the ESM loader):

```ts
const hoistedDir = path.join(workspaceRoot, 'node_modules', '.pnpm', 'node_modules');
```

Under the project-local virtual store, a package at `node_modules/.pnpm/<pkg>/node_modules/<pkg>` resolved undeclared requires through node's ancestor walk, which reaches **two** directories:

1. `node_modules/.pnpm/node_modules` — the hoisted dir, from the `node_modules/.pnpm` ancestor
2. `<workspaceRoot>/node_modules` — from the workspace root ancestor

Under the global virtual store the package's realpath is `<storeDir>/links/@/<pkg>/<version>/<hash>/node_modules/<pkg>`, with no workspace above it, so both are lost. The bridge replaces (1) and nothing replaces (2).

This matters because pnpm hoists only **non-direct** dependencies into `node_modules/.pnpm/node_modules`. Direct dependencies of the workspace root exist only in `<workspaceRoot>/node_modules`, so they are exactly the set the bridge cannot cover. Observed in the bit repo itself:

```
mocha          hoisted:no  root:yes
oxlint         hoisted:no  root:yes
cross-env      hoisted:no  root:yes
lodash.get     hoisted:yes root:no
```

## Reproduction

In a workspace with `enableGlobalVirtualStore: true`, have a store-resident package require a direct dependency of the workspace root that it does not declare itself. Concretely, in the bit repo (which enables it on the `enable-gvs` branch), mocha requiring its reporter:

```
$ ./node_modules/.bin/mocha --reporter mocha-multi-reporters ...
✖ ERROR: TypeError: Could not load reporter "mocha-multi-reporters":
 Error: Cannot find module '<workspace>/mocha-multi-reporters'
Require stack:
- <storeDir>/links/@/mocha/11.1.0/<hash>/node_modules/mocha/lib/cli/run-helpers.js
```

`mocha-multi-reporters` is a direct dependency of the workspace root. This particular invocation is not started by bit, so it gets no bridge at all — but the same require fails identically inside a bit-run process, because the bridge does not carry the workspace root's `node_modules`.

## Proposed fix

Add `<workspaceRoot>/node_modules` to the entries the bridge contributes, **after** the hoisted dir, matching the order of the original ancestor walk. Both halves already generalize over a list — the CommonJS side through `NODE_PATH`, and the ESM loader iterates `extraNodePaths` — so this is a matter of what `ensureHoistedDependencyResolution` puts on the path.

Points worth deciding in review:

- `ensureSelfInstallationBridge` uses the same function against the *installation* root, where `<root>/node_modules` is the installation's own package set. That is the behaviour the project-local layout had too, so it looks correct, but it widens what a bvm-installed bit can resolve and deserves a deliberate call.
- The ESM loader anchors `createRequire` inside each entry and relies on node not appending `/node_modules` to a directory already named `node_modules` — the comment at [hoisted-resolution-bridge.ts:70](https://github.com/teambit/bit/blob/master/scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts#L70) notes this holds exactly for entries named `node_modules`. The new entry is also named `node_modules`, so it stays in that case.

## Test

`e2e/harmony/global-virtual-store.e2e.ts` — a component or env that requires a workspace-root direct dependency it does not declare, asserted to resolve under the global virtual store.

---

# Second gap: TypeScript type resolution has no bridge at all

Everything above is about `require`/`import` at runtime. The same missing ancestor breaks `tsc`, and there the bridge does not exist in any form — `NODE_PATH` and the ESM loader are invisible to the TypeScript resolver.

## Symptom

A `.d.ts` in a store slot cannot walk up to the workspace's hoisted `@types`. Types do not error out visibly; they resolve to a different copy or collapse to `unknown`, and the failure surfaces as prop errors in whatever source the type graph reaches:

```
components/ui/code-compare/code-compare-editor-settings/code-compare-editor-settings.tsx(42,10):
error TS2322: Type '{ children: Element; className: string; checked: boolean; value: string; onInputChanged: () => void; }'
is not assignable to type 'IntrinsicAttributes & CheckboxLabelProps'.
  Property 'children' does not exist on type 'IntrinsicAttributes & CheckboxLabelProps'.
```

This repo already patches it by hand for its own `tsc`, in [tsconfig.json](https://github.com/teambit/bit/blob/master/tsconfig.json):

```jsonc
// Under pnpm's global virtual store, a published package's .d.ts can no longer reach the
// workspace's hoisted @types by walking up from its store slot (project-locally the walk
// out of node_modules/.pnpm found them). Pin the react typings to the workspace copies so
// type resolution matches the runtime NODE_PATH bridge.
"paths": {
  "react": ["./node_modules/@types/react"],
  "react-dom": ["./node_modules/@types/react-dom"],
  "@teambit/*": ["./node_modules/@teambit/*"]
}
```

That pin covers exactly two packages, in one hand-maintained file, for one workspace.

## Reproduction

Nothing but the pin separates working from broken, on the `enable-gvs` branch of this repo:

```
$ npx tsc --noEmit                       # as committed
0 errors

$ npx tsc --noEmit -p tsconfig.nopin.json   # same file, react/react-dom entries removed
components/ui/code-compare/code-compare-editor-settings/code-compare-editor-settings.tsx(42,10): error TS2322: ...
```

The build path has no pin, so it fails where the workspace succeeds. `e2e/harmony/extensions-config-diff.e2e.ts` fails in its `before` hook: `bit tag --build` runs `teambit.compilation/compiler:TSCompiler` for env `teambit.harmony/aspect` over `my-scope/ext1`, whose `@teambit/*` `.d.ts` files reach the repo's UI component sources, which reach for `react` typings from a store slot and get the wrong copy. Every `children`/`className` prop then fails to type-check.

Reproduced locally on `enable-gvs`, matching CI exactly — same hook, same task, same component, same first error, same total:

```
$ npm_config_bit_bin=bbit npx mocha --require ./babel-register --exit \
    e2e/harmony/extensions-config-diff.e2e.ts

  1) extensions config diff
       "before all" hook in "extensions config diff":
     Error: Command failed: bbit tag   --build
Failed task 1: "teambit.compilation/compiler:TSCompiler" of env "teambit.harmony/aspect"
component: my-scope/ext1@0.0.1
...
Found 188 errors in 1 components
✖ Total 15 tasks. 14 succeeded. 1 failed. 0 skipped. Total errors: 188.
```

The 188 are 47 distinct errors across 27 files, every file inside the workspace, the set repeated as the build prints it through the task result, the `BitError`, and the wrapping command failure.

## Proposed fix

Have the TypeScript compiler inject the type-resolution equivalent of the runtime bridge — `paths` / `typeRoots` pointing at the workspace's `node_modules` and its hoisted directory — whenever the workspace is on the global virtual store, so a capsule build resolves types the way the bridged process resolves requires. Deriving it from the same layout check (`isGlobalVirtualStoreLayout`) keeps the two halves from drifting.

With that in place, the hand-written `paths` block in this repo's `tsconfig.json` should be reviewed: it exists only because the resolver had no bridge, and it covers a set of packages that was never meant to be exhaustive.

## Test

`e2e/harmony/extensions-config-diff.e2e.ts` already fails on the `enable-gvs` branch and passes once types resolve. A dedicated case in `e2e/harmony/global-virtual-store.e2e.ts` — building a component whose type graph reaches a store slot — would pin the behaviour where the other global-virtual-store cases live.

### PR DESCRIPTION
**Pull request by zkochan**
Closes #10588.

Under pnpm's global virtual store a package's real directory sits outside the project, so the ancestor walk that used to satisfy its undeclared requires reaches nothing. The bridge in `hoisted-resolution-bridge.ts` restored one of the two directories that walk passed through, and only for `require` — not for `tsc`. This restores both, on both resolvers.

## Both directories, not just the hoisted one

pnpm hoists only **non-direct** dependencies into `node_modules/.pnpm/node_modules`. A direct dependency of the root lives in the root's own `node_modules` and nowhere else, so the walk out of a store slot reached two directories with disjoint contents and the bridge covered one. Measured in this repo:

```
mocha          hoisted:no  root:yes
oxlint         hoisted:no  root:yes
cross-env      hoisted:no  root:yes
lodash.get     hoisted:yes root:no
```

`hoistedResolutionDirs(root)` now returns both, in walk order, and `ensureHoistedDependencyResolution` puts both on `NODE_PATH` — the hoisted one still first, so it keeps winning exactly as the walk had it. The ESM loader mirrors the entry list, so it follows automatically.

## The type-resolution half

TypeScript reads neither `NODE_PATH` nor the ESM loader, so under this layout a `.d.ts` in a store slot resolves `react` to `react/index.js`, gets no typings, and every type derived from it degrades quietly. `--traceResolution` on this repo:

```
Resolving 'react' from STORE/@teambit/evangelist.input.checkbox.label/.../checkbox-label.d.ts
  → STORE/@/react/19.2.7/.../react/index.js          # untyped

Resolving 'react' from components/ui/.../code-compare-editor-settings.tsx
  → STORE/@types/react/19.2.14/.../index.d.ts
```

The two copies never unify, and it surfaces as `Property 'children' does not exist on type 'IntrinsicAttributes & CheckboxLabelProps'` — a message that names a prop rather than the cause. This repo already patches it by hand in `tsconfig.json` for its own `tsc`, for two packages, in one file.

TypeScript's only lever is `paths`, which is a *redirect* where `NODE_PATH` is a *fallback*, so a blanket mapping does real damage. The mapping is therefore kept to what the walk actually used to find. Full-repo `tsc --noEmit`, one rule at a time:

| mapping | errors |
| --- | --- |
| every `@types` package from both directories | 135 — overrides `glob`, `react-router-dom` and everything else that ships its own modern types |
| `@types/x` only where `x` itself ships no typings | 9 |
| + `@teambit/*` from the hoisted directory **and** the root | 5 — the hoisted directory's older transitive copies win |
| + `@teambit/*` from the **root only** | **0** |

So: a `@types/x` mapping only when `x` ships no typings of its own — that is the case the walk-up existed for, since a self-typed `x` stopped the walk before it ever reached `@types/x` — and `@teambit/*` from the root alone, because core aspects have to be the single copy from the running installation (the invariant `DependencyLinker` already maintains).

Applied in `TypescriptMain.createCompiler` after every transformer, so anything configured deliberately keeps its mapping and the bridge only fills in what would otherwise resolve to nothing. Two roots, each gated on the layout its own last install recorded: the workspace, whose components are what gets compiled, and the running installation, whose core aspects the compiled program reaches through the links `DependencyLinker` writes. A bvm installation is project-local and stays out entirely.

## Verification

- `bit test` on both components: **119 passing**, 16 of them new.
- New e2e case, `global-virtual-store.e2e.ts` → "building an aspect": builds an aspect in a workspace with `enableGlobalVirtualStore: true`. Passes on this branch; with the TypeScript half reverted and the aspect recompiled it fails with 47 `error TS`, so it fails for the reason it exists.
- `e2e/harmony/extensions-config-diff.e2e.ts`, which is what surfaced this: **7 passing, 0 failing, 0 TypeScript errors**. On a global-virtual-store workspace before this change it failed in its `before` hook with `Failed task 1: "teambit.compilation/compiler:TSCompiler"` and 188 errors attributed to one component.
- `npm run lint` clean.

## Not included

- The hand-written `paths` block in this repo's `tsconfig.json` stays. `npm run lint` runs `tsc` directly against that file and never goes through bit's compiler, so those pins are still load-bearing; retiring them is a separate change with its own verification.
- No e2e for the runtime half. `hoistedResolutionDirs` is unit-tested, but exercising a phantom require of a root-direct dependency from a store slot needs a published fixture package that under-declares one, and a test that would pass with or without the change is worse than none. The type-side e2e covers the shared directory logic end to end.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scopes/typescript/typescript/global-virtual-store-type-paths.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    let manifest: { types?: unknown; typings?: unknown };
+    try {
+      manifest = JSON.parse(fs.readFileSync(path.join(packageDir, 'package.json'), 'utf8'));
+    } catch {
+      continue; // not in this directory - keep looking in the next
+    }
+    if (typeof manifest.types === 'string' || typeof manifest.typings === 'string') return true;
+    // the implicit form: no `types` field, but an index.d.ts beside the entry point
+    return fs.existsSync(path.join(packageDir, 'index.d.ts'));
+  }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/typescript/typescript/global-virtual-store-type-paths.ts:116]: Fixed in 2f834311e — this was a real bug. `shipsOwnTypes` now counts a `types` condition anywhere in a conditional `exports` map, `typesVersions`, and the declaration file the resolver infers from the entry point (`dist/index.js` → `dist/index.d.ts`, a directory entry point → `<dir>/index.d.ts`), alongside the fields it already read.

Worth stating why it matters, since it isn't symmetric: a false *positive* only skips a mapping that was probably unnecessary, but a false *negative* redirects the specifier to `@types` and hides the package's own, usually newer, declarations — the exact damage the conservative rule exists to avoid. Five unit cases added, including the `main: dist/index.js` + `dist/index.d.ts` layout you called out and a negative case (javascript beside the entry point, no declarations) so the detection can't drift into matching everything.

No change to the outcome on this repo: re-measuring through the shipped code path, the mapping is the same 149 `@types` entries plus `@teambit/*`, and full-repo `tsc --noEmit` stays at 0 errors. The fix is defensive against layouts this workspace happens not to contain.

File: scopes/typescript/typescript/typescript.main.runtime.ts [hunk 2]
```diff
@@ -136,11 +141,38 @@ export class TypescriptMain {
       TypescriptAspect.id,
       this.logger,
       afterMutationWithoutTsconfig,
-      afterMutation.raw.tsconfig,
+      this.bridgeTypeResolution(afterMutation.raw.tsconfig),
       tsModule as any
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/typescript/typescript/typescript.main.runtime.ts:145]: Measured before deciding: the full scan is **7.1 ms** for all 166 `@types` entries across both bridge directories on this repo, which is about as large a workspace as this will meet. `createCompiler` is called per env, not per component, so this is single-digit milliseconds a handful of times per command.

I'd rather not trade that for a cache. The invalidation key you suggest, `.modules.yaml` mtime, doesn't actually cover the input — that manifest records the layout, while the mapping is derived from the *contents* of the two `@types` directories, so anything that changes a package without rewriting the manifest would serve a stale mapping. Getting it right means watching both directories, and the failure mode when it's wrong is a silently wrong `paths` block, which is the class of bug this whole PR is about.

The comment in the code says it's uncached deliberately and why (an install can move a workspace onto the global virtual store mid-process). If it ever shows up in a profile, the honest fix is memoizing for the duration of a single build rather than across a process.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts [hunk 3]
```diff
@@ -197,17 +205,37 @@ export function ensureSelfInstallationBridge(): void {
...
-  if (!fs.existsSync(hoistedDir)) return;
+  const dirs = hoistedResolutionDirs(workspaceRoot);
+  if (!dirs.length) return;
   const existing = process.env.NODE_PATH;
-  if (!existing?.split(path.delimiter).includes(hoistedDir)) {
-    process.env.NODE_PATH = existing ? `${hoistedDir}${path.delimiter}${existing}` : hoistedDir;
+  const known = new Set(existing?.split(path.delimiter).filter(Boolean) ?? []);
+  const added = dirs.filter((dir) => !known.has(dir));
+  if (added.length) {
+    // prepended in walk order, so the hoisted directory keeps winning over the root's own
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts]: Good catch — fixed in d539648e0. `NODE_PATH` is now rebuilt with both entries at the front in walk order, with everything the bridge doesn't own kept behind them in its existing relative order, and the write is skipped when the result already matches.

The reachable case isn't a hand-edited `NODE_PATH`, it's version skew: a bit whose bridge added the hoisted directory alone (every build before this PR) exports that to its children, and the child running this code would then add the root's `node_modules` *in front* of it — inverting exactly the walk the bridge exists to reproduce. Membership was the wrong idempotency test for a variable where position is the semantics.

Five unit cases added around the ordering. Verified the way the rest of this PR was: against the previous implementation exactly one of them fails, "should reorder entries a previous bridge left in the wrong order", and the other four pass either way as characterization.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.spec.ts [hunk 4]
```diff
@@ -36,3 +43,88 @@ describe('parseRecordedVirtualStoreDir()', () => {
...
+    fs.ensureDirSync(hoisted());
+    nodePath = process.env.NODE_PATH;
+    nodeOptions = process.env.NODE_OPTIONS;
+  });
+  afterEach(() => {
+    if (nodePath === undefined) delete process.env.NODE_PATH;
+    else process.env.NODE_PATH = nodePath;
+    if (nodeOptions === undefined) delete process.env.NODE_OPTIONS;
+    else process.env.NODE_OPTIONS = nodeOptions;
+    fs.removeSync(root);
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.spec.ts:105]: Fair — fixed in fa00cbdde. `afterEach` now rederives the module paths from the restored `NODE_PATH`, so the resolver stops pointing into the temp directory the very next line deletes, and `register` is taken away for the duration of each case so the irreversible half never runs. Removing `register` rather than stubbing it a no-op means the function early-returns through the guard it already has for runtimes without `module.register`, instead of running the body and mutating its own bookkeeping.

Worth noting the blast radius was small — the loader only engages after default resolution has already failed, and it would then have searched deleted directories and rethrown the original error — but five cases each chaining another permanent loader onto the test process is not something to leave in, and the stale `globalPaths` was a genuine ordering hazard for whatever ran next.

Verified the interop does what it needs to: the spec's `import Module from 'module'` compiles to the same object the source's `require('module')` returns, so the stub is visible at the call site. 79 tests passing, lint clean.

File: scopes/typescript/typescript/global-virtual-store-type-paths.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+ * without saying so: `dist/index.js` is typed by `dist/index.d.ts`, and a directory entry point by
+ * an `index.d.ts` inside it.
+ */
+function declarationsBesideEntry(packageDir: string, main: unknown): boolean {
+  const entry = typeof main === 'string' && main ? main : 'index.js';
+  const resolved = path.join(packageDir, entry);
+  const withoutExtension = resolved.replace(/\.(js|cjs|mjs|jsx)$/, '');
+  return (
+    fs.existsSync(`${withoutExtension}.d.ts`) ||
+    fs.existsSync(path.join(resolved, 'index.d.ts')) ||
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/typescript/typescript/global-virtual-store-type-paths.ts]: Right on both counts — fixed in a9abe6a5f, including the part marked optional.

`.d.mts`/`.d.cts` are now tried alongside `.d.ts` against each of the three shapes (sibling of the entry, `index.*` inside a directory entry, `index.*` at the package root). I deliberately try all three extensions everywhere rather than pairing each one with the entry's own extension: being too eager only skips a mapping, while missing one buries real declarations, and the asymmetry should decide the design rather than tidiness.

The `exports` hardening turned out to be worth doing rather than deferring. Under `moduleResolution: node` TypeScript ignores `exports` and reads `main`, which was already covered — but under `node16`/`nodenext` it resolves the export target and infers declarations from *that* path, so a package with `exports: { ".": { "import": "./dist/index.mjs" } }`, no `types` condition and no `main` was still a false negative. `shipsOwnTypes` now checks declarations beside every string target in the `exports` map at any depth, in addition to `main`. Test added for exactly that shape.

Outcome on this repo is unchanged again: 149 `@types` mappings, `react` still among them, full-repo `tsc --noEmit` at 0 errors. 53 unit tests, lint clean. Like the previous detection fix this is protection against layouts this workspace doesn't happen to contain.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts [hunk 6]
```diff
@@ -197,17 +205,40 @@ export function ensureSelfInstallationBridge(): void {
...
-  if (!fs.existsSync(hoistedDir)) return;
+  const dirs = hoistedResolutionDirs(workspaceRoot);
+  if (!dirs.length) return;
   const existing = process.env.NODE_PATH;
-  if (!existing?.split(path.delimiter).includes(hoistedDir)) {
-    process.env.NODE_PATH = existing ? `${hoistedDir}${path.delimiter}${existing}` : hoistedDir;
+  const untouched = (existing?.split(path.delimiter).filter(Boolean) ?? []).filter((dir) => !dirs.includes(dir));
+  // rebuilt rather than prepended, because order is resolution order and an entry already present
+  // is not necessarily in front of the one it has to beat: a bit that bridged the hoisted
+  // directory alone leaves it in `NODE_PATH` for its children, and adding the root's node_modules
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts]: Fixed in f2450a096 — entries are now matched by path rather than by spelling, so one naming a directory the bridge owns is replaced by the canonical entry instead of being kept beside it.

Rather than trimming separators ad hoc I reused the comparison this module already reasons about for containment: `path.relative(a, b) === ''` normalizes both sides, which covers a trailing separator, a redundant `.` segment, a relative spelling, and — on Windows, where it matters most — a difference in case. Same platform argument as `isPathInsideOrEqual` right above it, so the two stay consistent.

One correction to the framing, for whoever reads this later: the duplicate could not have changed which package won. The rebuilt list puts the canonical entries first and everything else behind them, so a differently-spelled duplicate would have sat after the entry it duplicates and only ever cost a redundant stat on a miss. The reason to fix it is that a `NODE_PATH` carrying the same directory twice in two spellings is hard to read when something does go wrong, not that resolution was wrong.

Four unit cases added, three of them on the comparison itself. Against the previous implementation the spelling case fails, as it should. 83 tests passing, lint clean.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts [hunk 7]
```diff
@@ -197,17 +215,44 @@ export function ensureSelfInstallationBridge(): void {
...
+ * walk reached them: the hoisted `node_modules/.pnpm/node_modules` first, then the root's own
+ * `node_modules`. Both are lost under the global virtual store, and they hold disjoint sets -
+ * pnpm hoists only *non-direct* dependencies, so a direct dependency of the root exists solely in
+ * the second. Missing directories are dropped; a root with neither yields an empty list.
+ *
+ * Shared with the type-resolution side of the bridge, which has to reach the same two directories
+ * through the mechanism TypeScript offers instead of `NODE_PATH`.
+ */
+export function hoistedResolutionDirs(root: string): string[] {
+  return [path.join(root, 'node_modules', '.pnpm', 'node_modules'), path.join(root, 'node_modules')].filter((dir) =>
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts:237]: This one doesn't reproduce. `npm run prettier:check` passes on this file, and so does `prettier --check` run directly against it — the repo's `printWidth` is 120 and both cited hunks are formatted exactly as prettier emits them; I ran `prettier --write` over every file in this PR after each change and it reported them unchanged.

There is one line over 120 characters in the file, line 298, but prettier accepts it because it is a single template literal it cannot break — and it is not from this PR: it sits unchanged on `master` at line 253, well before this branch.

Happy to reformat anything specific if you can point at a hunk where `prettier --check` actually disagrees.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts [hunk 8]
```diff
@@ -197,17 +215,44 @@ export function ensureSelfInstallationBridge(): void {
...
+    fs.existsSync(dir)
+  );
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts:239]: Keeping this as it is, but the concern is worth answering properly because part of it is real and part of it isn't.

**It cannot resolve a different copy.** `NODE_PATH` becomes `Module.globalPaths`, which node consults *after* the whole ancestor walk, so a bridge entry can only satisfy a require that would otherwise have thrown. Verified rather than recalled:

```
proj/node_modules/pkg  +  NODE_PATH=global (also holds pkg)   → LOCAL
proj/node_modules/pkg removed                                 → NODE_PATH
```

So the failure mode is masking a *missing* dependency, never shadowing a present one.

**Masking is real, but it is neither new nor avoidable here.** `ensureSelfInstallationBridge` already put the installation's hoisted directory on this same process-global variable before this PR, with the same visibility to workspace code; `NODE_PATH` has no notion of scope, so any self-installation bridging leaks by construction. Ordering is already the way round you'd want — the workspace bridge runs later and rebuilds the list with its own two directories in front, so the installation's entries are strictly the deeper fallback.

**Dropping the root for the self-installation case would make the bridge less faithful, not safer.** Under the project-local layout a package in the installation's `.pnpm` walked up into the installation's own `node_modules`, which is where its direct dependencies — including its `@teambit/*` copies — live. Restoring only the hoisted half would leave installation-side phantom requires that used to work now failing, which is the same asymmetry the type-side mapping is built around: over-broad costs a redundant last-resort lookup, too narrow breaks resolution outright. The module's stated design is to restore what the walk reached rather than a hand-picked subset.

Worth adding that the gate keeps this narrow: `ensureSelfInstallationBridge` only fires when the *installation itself* is linked to the global virtual store, which a bvm-installed bit never is. In practice that means bit running from source in a global-virtual-store workspace.

File: scopes/typescript/typescript/global-virtual-store-type-paths.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    return entries.some((entry) => declarationsBesideEntry(packageDir, entry));
+  }
+  return false;
+}
+
+/**
+ * `paths` entries that let a store slot resolve the types it used to reach by walking up out of
+ * `node_modules/.pnpm`. Empty for a root that is not on the global virtual store - callers gate on
+ * `isGlobalVirtualStoreLayout` - or one whose directories no longer exist.
+ *
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/typescript/typescript/global-virtual-store-type-paths.ts:130]: Fair — the doc claimed something the function has no way of knowing. Reworded in f5d7435c5 to say what it actually does: the mapping is derived from `hoistedResolutionDirs` alone, and the layout check belongs to the caller.

I took the documentation option rather than moving `isGlobalVirtualStoreLayout` inside, for two reasons. The caller has to make that decision per root regardless — it holds both the workspace and the installation root and gates each one separately — so an internal check would re-read `.modules.yaml` to re-derive a decision that was already made. And the mapping is not actually harmful for a project-local root: it would point at the same `@types` directories that root's own walk already reaches. Unnecessary rather than wrong, which is exactly why the guard belongs where the choice is, not buried in the deriving function.

The new wording says all of that, so the next call site has the contract in front of it.

File: scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts [hunk 10]
```diff
@@ -197,17 +215,44 @@ export function ensureSelfInstallationBridge(): void {
...
+  // rebuilt rather than prepended, because order is resolution order and an entry already present
+  // is not necessarily in front of the one it has to beat: a bit that bridged the hoisted
+  // directory alone leaves it in `NODE_PATH` for its children, and adding the root's node_modules
+  // in front of it there would invert the walk the bridge exists to reproduce. Everything this
+  // bridge does not own keeps its relative order, behind what the walk reached first.
+  const next = [...dirs, ...untouched].join(path.delimiter);
+  if (next !== existing) {
+    process.env.NODE_PATH = next;
     // `NODE_PATH` is read once when the module system initializes, so a later assignment only
```
[the remark this answers is not in the document]
Comment (zkochan) [reply scopes/dependencies/dependency-resolver/hoisted-resolution-bridge.ts:265]: Real, and fixed in 82540b435 — the skip test asked whether the entry *set* had grown, which a reorder leaves untouched, while the loader inlines its list and searches it in order. It now tracks the ordered list, so a reordering re-registers.

One correction to the premise, because it changes what the fix buys. I checked how node chains these rather than assuming:

```
register(FIRST); register(SECOND); import('./target.mjs')
→ enter SECOND, enter FIRST, exit FIRST, exit SECOND
```

The *last* registered loader runs first, and ours delegates through `defaultResolve` before reaching its own fallback — so an earlier registration resolves anything it can before a later one is ever consulted. Re-registering therefore cannot correct in-process precedence; the earlier loader keeps it for the entries it already had. What it does correct is the `--import` flag children inherit, which is replaced rather than stacked, and that was genuinely contradicting the `NODE_PATH` sitting beside it. That is the concrete divergence here, and it is now covered by a test that fails against the previous implementation.

The in-process limit is inherent to inlining the list, which the loader has to do — the hooks run on a thread whose `process.env` is a snapshot from registration time, so it cannot read `NODE_PATH` itself. Removing the limit means a single registration fed by a `MessagePort`, which is a larger change than this PR should carry; the bound is narrow (a specifier resolvable from more than one entry, across two registrations with different orders, needing a process that bridges two roots in alternating order) and the CommonJS half is rebuilt on every call and always right. I have written that trade-off into the module docs, including the corrected chaining behaviour, since the previous comment there claimed the older loader chained "harmlessly" — true for a subset, not for a reorder.

85 tests passing, lint clean.

**Review by GiladShoham — APPROVED** [review]
